주문 결제 발행 api 구현

wanniDev opened the pull request:
카프카 학습에 집중하기 위해 프로젝트 작업의 경우, 
기존 코드 리팩토링 및 멱등키 생성에 사용되는 결제 주문(paymentOrder) 발행 api를 구현해보았습니다.

시간이 허락된다면, 간단한 퍼블리싱을 적용하여, 직접 주문 결제(상품 상세 페이지에서 옵션을 정하고, '바로 결제하기'를 선택하는 상황)를
수행할 수 있는 기능도 추가하고 merge 해보겠습니다.

wanniDev committed:
Refactor: 보관기간 옵션 제거

결제 기록 추적을 위해 우선은 로그를 별도로 삭제하지는 않는 방향 유지

wanniDev committed:
Refactor: 결제 승인전 주문 결제 검증하는 validator를 좀 더 가독성있게 함수로 래핑

wanniDev committed:
Feat: paymentOrder 발급 api 시그니처 구현

1. 사용자가 물건을 고른다.
2. 장바구니 페이지를 통해 '구매하기'를 클릭하거나 상세 페이지에서 '바로 구매하기'를 클릭한다. (사용자의 구매 확정시점)
3. 해당 상품의 구매정보(현 시점에선 구매 상품 주문 명, 구매 수량)를 바탕으로 발급번호 발행
4. 발행한 발급번호를 토대로 주문 결제를 수행하는 uri로 리다이렉트

wanniDev committed:
Feat: 주문 결제 발행 비즈니스 로직 구현

1. paymentService에서 주문 결제 command를 처리해주는 paymentOrderProcessor에게 주문 결제 발행 위임
2. paymentOrderProcessor는 command 객체의 파라미터 값을 검증하고 새로운 주문 결제 발행 후 해당 주문 결제의 id 리턴
3. 그 id를 기준으로 결제를 진행할 수 있는 주문 결제 페이지로 리다이렉트

* PRG(Post Redirect Get)를 꼭 써야할까? 그냥 GET으로 발행과 동시에 페이지 렌더링이 가능할 수 있다. 하지만, 뒤로가기와 같이 사용자의 오류에 의한 중복 발행 이슈를 대응하기 위한 새로운 로직이 필요하다.
PRG를 활용하면, 별도의 로직 구성이 없어도 중복 발행 이슈를 검증된 방식으로 대응할 수 있다.
* 이메일 검증의 경우, 사용자가 본인 이메일 계정의 메일 내부에 작성된 검증 url 하나만으로 수행해야하기 때문에 Post를 수행하는게 불가능하다. 하지만, Get으로 검증 url로 이메일 검증을 시도하고, 다시 로그인 페이지로 리다이렉트 시키는 방식을 활용해볼 수 있다.

wanniDev committed:
Test: 기능 명세를 위해 결제 주문 발행 api 테스트 추가

f-lab-troy commented on payment-api/payment-application/src/main/kotlin/org/collaborators/paymentslab/payment/application/PaymentService.kt:
```diff
@@ -38,4 +41,8 @@ class PaymentService(
         return entities.map { PaymentHistoryQueryQueryModel.of(it) }
     }
+
+    fun generatePaymentOrder(command: PaymentOrderCommand): String {
```
지난 번에 이야기 나눈대로 명확한 naming도 생각해보시면 좋습니다.
- generateXXX -> generator
- processXXX - processor

wanniDev replied:
수정 완료하였습니다. 아울러 해당 부분의 기능을 구현하던 중, 어쩌면 이벤트 스토어 전용 DB를 구축할 필요없이 카프카 스트림즈를 활용해도 될 것 같다는 생각이 들었습니다. 괜찮으시다면, 카프카 스터디가 마무리 된 이후에 한번 의논해보았으면 합니다..!

f-lab-troy commented on payment-api/payment-infrastructure/src/main/kotlin/org/collaborators/paymentslab/payment/infrastructure/tosspayments/TossPaymentsValidator.kt:
```diff
@@ -6,41 +6,46 @@ import org.collaborators.paymentslab.payment.domain.entity.PaymentsStatus
 import org.collaborators.paymentslab.payment.infrastructure.tosspayments.exception.*
 import org.slf4j.LoggerFactory
 import org.springframework.security.core.context.SecurityContextHolder
+import java.util.*
```
(*)로 패키지 전체를 import 하는 것과 사용되는 class만 import하는 것의 차이점을 인지하면 좋을 것 같습니다.

wanniDev replied:
무분별하게 와일드 카드를 사용하여 패키지 내의 모든 클래스를 imort를 할 경우, 가독성 저하 및 의존성의 불명확성, 그리고 컴파일 시간 증가 등의 불이익이 있으니 주의하면 좋을 거 같다로 이해했습니다.

왜인지는 모르지만, paymentOrder의 status를 소문자로 변형시키는 과정에서 사용되는 `Locale.getDefault()`를 사용할 때마다 인텔리제이가 `import java.util.*`를 자동으로 추가해주네요. 해당 부분은  제가 인위적으로 `import java.util.Locale`로 다시 수정했습니다.

f-lab-troy commented on payment-api/payment-infrastructure/src/main/kotlin/org/collaborators/paymentslab/payment/infrastructure/tosspayments/TossPaymentsValidator.kt:
```diff
@@ -6,41 +6,46 @@ import org.collaborators.paymentslab.payment.domain.entity.PaymentsStatus
-            log.error("invalid amount from paymentOrderId {}", paymentOrderId)
-            throw InvalidPaymentOrderException()
-        }
+        checkAccountId(paymentOrder, accountUser.id)
```
지금은 일단 간단한 리팩토링만 하신 것으로 이해했습니다.
세션에서 이야기 나눈대로 추후에는 확장성을 위해
TossPaymentsValidator 인터페이스를 구현하는 AccountIdValidator, PaymentDetailsValidator등으로 나누어 처리하는 것도 고려해보면 좋을 것 같습니다.

f-lab-troy commented on payment-api/payment-presentation/src/main/kotlin/org/collaborators/paymentslab/payment/presentation/PaymentApi.kt:
```diff
@@ -4,10 +4,15 @@ import jakarta.validation.Valid
+import java.net.URI
 
 @RestController
 @RequestMapping("api/v1/toss-payments")
```
보통 버전 (v1)은 subdomain (eg. v1.api.domain.com/...) 또는 path의 가장 앞에 (eg. api.domain.com/v1/...) 쓰는게 일반적입니다.

wanniDev replied:
요즘에 localhost 환경에서만 작업을 하나보니, 이 부분을 미처 고려하지 못했습니다.
다만, 결제 관련 api를 직접 시연해볼 html 페이지 부분도 일부 추가할  예정인점을 감안하여, 우선 기본 url은 상수로 분류하고 path에는 'api'를 넣어놓겠습니다.

만약, 따로 렌더링 서버를 구축할 예정이 생긴다면, path에 api를 완전히 없애는 방향으로 진행 해보겠습니다.

wanniDev committed:
Refactor: api 시그니처 버전 내용 변경

기존에 호스트의 도메인에 api를 포함하여 해당 호스트가 api 서빙용으로만 사용된다는 것을 명시적으로 나타낸다면,
경로에 'api'를 따로 작성할 필요는 없다.

wanniDev committed:
Fix: 불필요한 전체 라이브러리 패키지 import 구문 수정

wanniDev committed:
Fix: 하드코딩으로 입력된 테스트 코드의 api 시그니처 내용 수정

wanniDev committed:
Feat: 결제주문 발행 테스트를 시연해보기 위해 구매 페이지 html 매핑

wanniDev committed:
Fix: 최초 주문 발행 시점에 accountId를 검사하는 코드는 잠시 비활성화

브라우저단에서는 얼마든지 주문 발행자의 계정 정보를 얼마든지 변조할 수 있기 때문에, 백엔드단에서 한번 더 검증하는 것이 맞다.
하지만, 현재는 결제를 위한 테스트를 그냥 html/css/javascript를 활용하고 있고, 그 상태에서는 jwt 라이브러리를 활용하기가 어렵기 때문에
일단은 결제 주문 발행 flow가 의도대로 흘러가는지만 확인하고 프론트에 jwt를 디코딩하는 코드를 추가한 다음에 다시 활성화 하기로 했다.

wanniDev committed:
Fix: 어차피 원복시킬 부분이니, 기존 결제 주문 발행 기능의 테스트가 통과할 정도로만 활성화

wanniDev committed:
Fix: 주문발행에 활용될 accountId는 백엔드 코드는 건들지 않고 html에 임시로 하드코딩하는 방향으로 변경

wanniDev committed:
Refactor: api를 호출하는 함수의 기능 명세를 더 명확하게 하기 위해, 네이밍 변경 및 application 코드 분리
